testing-internal/src/main/java/com/linecorp/armeria/internal/testing/AbstractMethodChainingTest.java

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.testing;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+import org.reflections.ReflectionUtils;
+import org.reflections.Reflections;
+import org.reflections.adapters.JavaReflectionAdapter;
+import org.reflections.scanners.SubTypesScanner;
+import org.reflections.util.ClasspathHelper;
+import org.reflections.util.ConfigurationBuilder;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Makes sure most builder overrides all overridden methods with the correct return type.
+ */
+public abstract class AbstractMethodChainingTest {
+
+    @Test
+    void methodChaining() {
+        final String packageName = getClass().getPackage().getName();
+        findAllClasses(packageName).stream()
+                                   .map(ReflectionUtils::forName)
+                                   .filter(AbstractMethodChainingTest::filterClass)
+                                   .forEach(clazz -> {
+                                       final List<Method> methods = obtainMethods(clazz);
+                                       for (Method m : methods) {
+                                           try {
+                                               final Method overriddenMethod =
+                                                       clazz.getDeclaredMethod(m.getName(),
+                                                                               m.getParameterTypes());
+                                               assertThat(overriddenMethod.getReturnType()).isSameAs(clazz);
+                                           } catch (NoSuchMethodException e) {
+                                               // ignored
+                                           }
+                                       }
+                                   });
+    }
+
+    private static Collection<String> findAllClasses(String packageName) {
+        final ConfigurationBuilder configuration = new ConfigurationBuilder()
+                .setUrls(ClasspathHelper.forPackage(packageName))
+                .setScanners(new SubTypesScanner())
+                .setMetadataAdapter(new JavaReflectionAdapter());
+        final Reflections reflections = new Reflections(configuration);
+        return reflections.getStore().get(SubTypesScanner.class.getSimpleName()).values();
+    }
+
+    private static boolean filterClass(Class<?> clazz) {
+        return declaredInTestClass(clazz) &&
```

Comment:
We don't need to call this method because we always check if the name ends with "Builder"?
(the method check if the name of the class ends with "Test" which is exclusive)

Reply:
Removed